src/ert/_c_wrappers/enkf/enkf_main.py

```diff
@@ -362,47 +473,74 @@ def sample_prior(
+                    props = xtgeo.gridproperty_from_file(
+                        init_file, name=parameter, grid=grid
+                    )
+
+                    data = props.values1d.data
+                    field_config = config_node.getFieldModelConfig()
+                    trans = field_config.get_init_transform_name()
+                    data_transformed = field_config.transform(trans, data)
+
+                    storage.save_field_data(parameter, realization_nr, data_transformed)
+
+            elif impl_type == ErtImplType.GEN_KW:
+                gen_kw_config = config_node.getKeywordModelConfig()
+                keys = list(gen_kw_config)
                 if config_node.get_init_file_fmt():
-                    parameter_values = gen_kw_node.values_from_files(
+                    parameter_values = gen_kw_config.values_from_files(
                         active_realizations,
                         config_node.get_init_file_fmt(),
                         keys,
                     )
                 else:
-                    parameter_values = gen_kw_node.sample_values(
+                    parameter_values = gen_kw_config.sample_values(
                         parameter,
                         keys,
                         str(self._global_seed.entropy),
                         active_realizations,
                         self.getEnsembleSize(),
                     )
-                storage.save_parameters(
-                    config_node,
-                    active_realizations,
-                    _clib.update.Parameter(parameter),
-                    parameter_values,
+
+                storage.save_gen_kw(
+                    parameter_name=parameter,
+                    parameter_keys=keys,
+                    parameter_transfer_functions=gen_kw_config.get_priors(),
+                    realizations=active_realizations,
+                    data=parameter_values,
                 )
-            else:
+            elif impl_type == ErtImplType.SURFACE:
                 for realization_nr in active_realizations:
-                    _clib.enkf_state.state_initialize(
-                        enkf_node,
-                        storage,
-                        realization_nr,
+                    init_file = config_node.get_init_file_fmt()
+                    if "%d" in init_file:
+                        init_file = init_file % realization_nr
+                    storage.save_surface_file(
+                        config_node.getKey(), realization_nr, init_file
                     )
+            else:
+                raise NotImplementedError(f"{impl_type} is not supported")
         for realization_nr in active_realizations:
-            if state_map[realization_nr] in [STATE_UNDEFINED, STATE_LOAD_FAILURE]:
-                state_map[realization_nr] = RealizationStateEnum.STATE_INITIALIZED
+            storage.update_realization_state(
```

Comment:
`save_field_data`, `save_gen_kw` and `save_surface_file` all change realization state from `STATE_UNDEFINED` to `STATE_INITIALIZED`. Why do we here want to change state from `STATE_UNDEFINED` and `STATE_LOAD_FAILURE` to `STATE_INITIALIZED`?

Reply:
it is old behaviour. I guess it is to make it possible for re-running failed realizations

Reply:
How do we know that realization failed when its state gets set to `STATE_INITIALIZED` if it has state `STATE_LOAD_FAILURE`?

Reply:
Overwriting the state STATE_LOAD_FAILURE only happens if you are writing parameters to an existing case, essentially saying it is now good to go for a new evaluation. The user should now whether the specific realization failed prior to rerunning a case, but once the rerun is initialized the information is lost if that specific realization failed in a previous evaluation.